modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1113 @@
+
+  // if the node is outside of the structural model, store its position in _boundary to
+  // _inactive_boundary_pos
+  for (unsigned int j = 0; j < n_boundary; ++j)
+  {
+    Node * this_node = _cut_mesh->node_ptr(_boundary[j]);
+    mooseAssert(this_node, "Node is NULL");
+    Point & this_point = *this_node;
+
+    const Elem * elem = (*pl)(this_point);
+    if (elem == NULL)
+      _inactive_boundary_pos.push_back(j);
+  }
+
+  unsigned int n_inactive_boundary = _inactive_boundary_pos.size();
+
+  // all nodes are inactive, stop
+  if (n_inactive_boundary == n_boundary)
+    _stop = 1;
+
+  // find and store active boundary segments in "_active_boundary"
+  if (n_inactive_boundary == 0)
+    _active_boundary.push_back(_boundary);
+  else
+  {
+    for (unsigned int i = 0; i < n_inactive_boundary - 1; ++i)
+    {
+      if (_inactive_boundary_pos[i + 1] - _inactive_boundary_pos[i] != 1)
+      {
+        std::vector<BoundaryID> temp;
+        for (unsigned int j = _inactive_boundary_pos[i]; j <= _inactive_boundary_pos[i + 1]; ++j)
+        {
+          temp.push_back(_boundary[j]);
+        }
+        _active_boundary.push_back(temp);
+      }
+    }
+    if (_inactive_boundary_pos[n_inactive_boundary - 1] - _inactive_boundary_pos[0] <
+        n_boundary - 1)
+    {
+      std::vector<BoundaryID> temp;
+      for (unsigned int j = _inactive_boundary_pos[n_inactive_boundary - 1]; j < n_boundary; ++j)
+        temp.push_back(_boundary[j]);
+      for (unsigned int j = 0; j <= _inactive_boundary_pos[0]; ++j)
+        temp.push_back(_boundary[j]);
+      _active_boundary.push_back(temp);
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::findActiveBoundaryDirection()
+// find growth direction of each boundary node; this is currently assgined by a hard-coded growth
+// function;
+// it will be updated to (1) growth function specified in the input file,
+// and/or (2) growth direction determined by fracture mechanics
+{
+  _active_direction.clear();
+
+  for (unsigned int i = 0; i < _active_boundary.size(); ++i)
+  {
+    std::vector<Point> temp;
+    Point dir;
+
+    if (_inactive_boundary_pos.size() != 0)
+    {
+      for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+        dir(j) = 0;
+      temp.push_back(dir);
+    }
+
+    unsigned int i1 = 1;
+    unsigned int i2 = _active_boundary[i].size() - 1;
+    if (_inactive_boundary_pos.size() == 0)
+    {
+      i1 = 0;
+      i2 = _active_boundary[i].size();
+    }
+
+    for (unsigned int j = i1; j < i2; ++j)
+    {
+      Node * this_node = _cut_mesh->node_ptr(_active_boundary[i][j]);
+      mooseAssert(this_node, "Node is NULL");
+      Point & this_point = *this_node;
+      std::array<Real, LIBMESH_DIM> x;
+      for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+        x[j] = this_point(j);
+
+      // Here is the predefined crack growth direction
+      dir(0) = 5 * (x[0] - 0.3) + x[2];
+      dir(1) = 5 * (x[1] - 0.5) + (x[2] + x[0]) / 2;
+      dir(2) = 5 * (x[2] - 0.1) + x[0];
```

Comment:
I know that this is just a placeholder for doing something more general, but I hate having a hardcoded growth function here. You could supply references to three `Function` objects, that can supply the three components of the growth direction as a function of position and time. In the input file, those can just be parsed functions to do what you are doing here. That may not be the way we want to do this in the end, but at least it would be definable in the input file.